fw/http_parser.c

```diff
@@ -7504,6 +7998,409 @@ __h2_req_parse_x_forwarded_for(TfwHttpMsg *hm, unsigned char *data, size_t len,
+		if (unlikely(!__fsm_sz))
+			return CSTR_NEQ;
+		__FSM_H2_I_MOVE_fixup(Req_I_Forwarded_For_Sep, __fsm_sz, TFW_STR_VALUE);
+        }
+
+	/*
+	 * At this state we know that we saw at least one character as
+	 * a host address and now we can pass zero length token.
+	 */
+	__FSM_STATE(Req_I_Forwarded_For_Node_Id_Unquoted) {
+		__FSM_H2_I_MATCH_MOVE_fixup(xff, Req_I_Forwarded_For_Node_Id_Unquoted, TFW_STR_VALUE);
+		__FSM_H2_I_MOVE_fixup(Req_I_Forwarded_For_Sep, __fsm_sz, TFW_STR_VALUE);
+	}
+
+	__FSM_STATE(Req_I_Forwarded_For_Quoted) {
+		__FSM_H2_I_MATCH_MOVE_NEQ_fixup(xff, Req_I_Forwarded_For_Node_Id_Quoted, TFW_STR_VALUE);
+		if (unlikely(!__fsm_sz))
+			return CSTR_NEQ;
+		if (unlikely(*(p + __fsm_sz) != '"'))
+			return CSTR_NEQ;
+
+		__msg_hdr_chunk_fixup(p, __fsm_sz);
+		__msg_chunk_flags(TFW_STR_HDR_VALUE | TFW_STR_VALUE);
+		__FSM_H2_I_MOVE_nofixup_n(Req_I_Forwarded_For_Sep, __fsm_sz + 1);
+        }
+
+	__FSM_STATE(Req_I_Forwarded_For_Node_Id_Quoted) {
+		__FSM_H2_I_MATCH_MOVE_NEQ_fixup(xff, Req_I_Forwarded_For_Node_Id_Quoted, TFW_STR_VALUE);
+		if (unlikely(*(p + __fsm_sz) != '"'))
+			return CSTR_NEQ;
+
+		__msg_hdr_chunk_fixup(p, __fsm_sz);
+		__msg_chunk_flags(TFW_STR_HDR_VALUE | TFW_STR_VALUE);
+		__FSM_H2_I_MOVE_nofixup_n(Req_I_Forwarded_For_Sep, __fsm_sz + 1);
+	}
+
+	__FSM_STATE(Req_I_Forwarded_For_Sep) {
+		/* go to next param */
+		if (likely(c == ';'))
+			__FSM_H2_I_MOVE_fixup(Req_I_Forwarded, 1, 0);
+		/* Fall through */
+	}
+
+	__FSM_STATE(Req_I_Forwarded_For_Sep_End) {
+		/*
+		 * Proxy chains are rare, so we expect that the list will end
+		 * after the first node and we get EOL here.
+		 */
+		if (likely(IS_CRLF(c)))
+			return __data_off(p);
+		/* "for=" can be represented as comma separated list, find next one */
+		if (unlikely(c == ','))
+			__FSM_H2_I_MOVE_NEQ_fixup(Req_I_Forwarded_For_List, 1, 0);
+	        /* OWS before comma or before EOL (is unusual). */
+		if (unlikely(IS_WS(c)))
+			__FSM_H2_I_MOVE_fixup(Req_I_Forwarded_For_Sep_End, 1, 0);
+
+                return CSTR_NEQ;
+       }
+
+	__FSM_STATE(Req_I_Forwarded_Host_Start) {
+		if (likely(c == '"'))
+			__FSM_H2_I_MOVE_nofixup(Req_I_Forwarded_Host_Start_Quoted);
+		/* Fall through */
+	}
+
+	__FSM_STATE(Req_I_Forwarded_Host_Start_Unquoted) {
+		__fsm_sz = __data_off(p);
+
+		if (likely(isalnum(c) || c == '.' || c == '-'))
+			__FSM_I_JMP(Req_I_Forwarded_Host_Unquoted);
+		return CSTR_NEQ;
+	}
+
+	__FSM_STATE(Req_I_Forwarded_Host_Unquoted) {
+		/* See Req_UriAuthority processing. */
+		if (likely(isalnum(c) || c == '.' || c == '-'))
+			__FSM_H2_I_MOVE_LAMBDA_fixup_pos(Req_I_Forwarded_Host_Unquoted, 1,
+							 data + __fsm_sz,
+							 (p - (data + __fsm_sz)), {
+								__msg_hdr_chunk_fixup(data + __fsm_sz,
+										     (p - (data + __fsm_sz)));
+								__msg_chunk_flags(TFW_STR_HDR_VALUE
+										  | TFW_STR_VALUE);
+								__FSM_EXIT(CSTR_EQ);
+							 },
```

Comment:
This looks overcomplicated. `data + __fsm_sz = data + p - data = p`, so we can just use existing macros operating with `p`, i.e. `__FSM_H2_I_MOVE_LAMBDA_fixup_f()` - it fixes up the data chunk before `goto`, so we don't need to include the fixup logic into `lambda`.